src/ChannelServer/World/Inventory.cs

```diff
@@ -209,19 +220,62 @@ public Item GetItem(EquipSlot slot)
 		/// </summary>
 		/// <param name="item"></param>
 		/// <return>Index of the item.</return>
-		public int AddSilent(Item item)
+		public Dictionary<int, int> AddSilent(Item item)
 		{
 			var cat = item.Data.Category;
-
 			lock (_syncLock)
 			{
+				Dictionary<int, int> itemsUpdated = new Dictionary<int, int>();
+
 				if (!_items.ContainsKey(cat))
 					throw new ArgumentException("Unknown item category.");
 
-				_items[cat].Add(item);
-				_itemsWorldIndex[item.WorldId] = item;
+				var inventoryIndex = -1;
+				var expectedAmount = item.Amount;
+
+				if (item.Data.MaxStack > 1 && this.HasItem(item.Id))
+				{
+					int itemIndex = inventoryIndex;
+					while (expectedAmount > 0)
+					{
+						inventoryIndex = this.GetStackableItemIndex(cat, item.Id, item.Amount);
+						if (inventoryIndex != -1 && _items[cat].Count > inventoryIndex)
+						{
+							var spaceInStack = _items[cat][inventoryIndex].Data.MaxStack - _items[cat][inventoryIndex].Amount;
```

Comment:
Another tip, if you create a local variable for that item, the code becomes a little easier to read, and should you ever need to change something, you have less to do.

E.g. `var spaceInStack = item.Data.MaxStack - item.Amount;`
You see what's going on in an instant, without having to read the indices for the lists, to find out what you're working with.